<h1>Data Cleaning / Munging</h1>

<p>In these notes we look at creating various plots based on a set of data.  We have a datafile called degreeClassification.csv (available on blackboard) which consists of the following columns:</p>
<ul>
<li>Academic Year:  The year that the degree was awarded.</li>
<li>Mode of study: Full-time, part-time or both.</li>
<li>Sex: Male, Female, Other or all.</li>
<li>Country of HE provider: England, Northern Ireland, Scotland, Wales or all.</li>
<li>Classification of First Degree: First, Upper Second, Lower Second or Third/Pass.</li>
<li>Number: The number of students achieving that degree.</li>
<li>Percentage:  The percentage of students achieving that degree.</li>
</ul>   

<p>We start the exercises by getting the data ready to be plotted</p>

<h2>Import PANDAs</h2>

In [55]:
import pandas as pd

<h2>Read the data into a PANDAs DataFrame</h2>
<p>Read the 'degreeClassification.csv' in to a data frame called df, use the head method to output the first 20 rows of the datafile so that the structure of the dataframe can be explored.</p>

In [56]:
df = pd.read_csv("degreeClassification.csv")
df.head(20)

Title  \
0                   Location   
1             Academic years   
2                Data source   
3           Data source link   
4   Data file canonical link   
5                    Licence   
6                  Code page   
7                 Disclaimer   
8              2013/14 total   
9              2014/15 total   
10             2015/16 total   
11             2016/17 total   
12             2017/18 total   
13              Last updated   
14                       NaN   
15                       NaN   
16             Academic Year   
17                   2013/14   
18                   2013/14   
19                   2013/14   

   Percentage of first degree qualifiers obtaining each classification by sex, mode of study and country of HE provider  \
0                                                  UK                                                                     
1                                  2013/14 to 2017/18                                                                     
2                                                HESA                                                                     
3   https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
4   https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
5   Creative Commons Attribution 4.0 International...                                                                     
6                                      Unicode UTF-8                                                                      
7   Please note that this data includes rounded to...                                                                     
8                                              395605                                                                     
9                                              370905                                                                     
10                                             376330                                                                     
11                                             391875                                                                     
12                                             396940                                                                     
13                                             Jan-19                                                                     
14                                                NaN                                                                     
15                                                NaN                                                                     
16                                      Mode of study                                                                     
17                                                All                                                                     
18                                                All                                                                     
19                                                All                                                                     

   Unnamed: 2              Unnamed: 3                      Unnamed: 4  \
0         NaN                     NaN                             NaN   
1         NaN                     NaN                             NaN   
2         NaN                     NaN                             NaN   
3         NaN                     NaN                             NaN   
4         NaN                     NaN                             NaN   
5         NaN                     NaN                             NaN   
6         NaN                     NaN                             NaN   
7         NaN                     NaN                             NaN   
8         NaN                     NaN                             NaN   
9         NaN                     Na

<h2>Skipping Summary Rows</h2>
<p>Read the datafile and skip the rows that contain summary information.  Use the head method and output the first 20 rows.  The column names should be:
<ul>    
<li>Academic Year</li>
<li>Mode of study</li>
<li>Sex</li>
<li>Country of HE provider</li>
<li>Classification of First Degree</li>
<li>Number</li>
<li>Percentage</li>
</ul>    
</p>

In [57]:
df = pd.read_csv("degreeClassification.csv", skiprows=17)
df.head(20)

Academic Year Mode of study  Sex Country of HE provider  \
0        2013/14           All  All                    All   
1        2013/14           All  All                    All   
2        2013/14           All  All                    All   
3        2013/14           All  All                    All   
4        2013/14           All  All                    All   
5        2013/14           All  All                England   
6        2013/14           All  All                England   
7        2013/14           All  All                England   
8        2013/14           All  All                England   
9        2013/14           All  All                England   
10       2013/14           All  All       Northern Ireland   
11       2013/14           All  All       Northern Ireland   
12       2013/14           All  All       Northern Ireland   
13       2013/14           All  All       Northern Ireland   
14       2013/14           All  All       Northern Ireland   
15       2013/14           All  All               Scotland   
16       2013/14           All  All               Scotland   
17       2013/14           All  All               Scotland   
18       2013/14           All  All               Scotland   
19       2013/14           All  All               Scotland   

   Classification of First Degree  Number Percentage  
0                           First   79410        20%  
1                    Lower second   95965        24%  
2                      Third/Pass   21905         6%  
3                Total classified  395605        NaN  
4                    Upper second  198320        50%  
5                           First   68590        20%  
6                    Lower second   81665        24%  
7                      Third/Pass   19385         6%  
8                Total classified  339170        NaN  
9                    Upper second  169530        50%  
10                          First    1590        19%  
11                   Lower second    2215        26%  
12                     Third/Pass     285         3%  
13               Total classified    8560        NaN  
14                   Upper second    4470        52%  
15                          First    5125        21%  
16                   Lower second    5505        22%  
17                     Third/Pass     875         4%  
18               Total classified   24880        NaN  
19                   Upper second   13380        54%

<h2>Unique values in a Column</h2>
<p>Output the unique values for the column 'Mode of study' to give a general idea of the categorical data for this column</p>

In [58]:
df.columns
df['Mode of study'].unique()

array(['All', 'Full-time', 'Part-time'], dtype=object)

<h2>Boolean Masks</h2>

<p>Create a boolean mask and then filter the data based on the categorical data value 'All' for 'Mode of Study'.  We do this in stages to help understand how the more complex aggregated instruction is built.</p>
<p>The overall goal is to apply a mask so that all the rows that contain values part-time or full-time are dropped from the dataframe.  We do this so that we don’t double count, i.e. A count in 'All' will also also be a count in either 'Part-time' or 'Full-time'.</p>

<ol>
<li>Create a mask to output True/False for each row, True if 'Mode of study' is equal to 'All' and False otherwise</li>

In [59]:
df['Mode of study'] == "All"


0        True
1        True
2        True
3        True
4        True
5        True
6        True
7        True
8        True
9        True
10       True
11       True
12       True
13       True
14       True
15       True
16       True
17       True
18       True
19       True
20       True
21       True
22       True
23       True
24       True
25       True
26       True
27       True
28       True
29       True
        ...  
1350    False
1351    False
1352    False
1353    False
1354    False
1355    False
1356    False
1357    False
1358    False
1359    False
1360    False
1361    False
1362    False
1363    False
1364    False
1365    False
1366    False
1367    False
1368    False
1369    False
1370    False
1371    False
1372    False
1373    False
1374    False
1375    False
1376    False
1377    False
1378    False
1379    False
Name: Mode of study, Length: 1380, dtype: bool

<ol start='2'>
<li>Assign the Boolean mask to a variable named mask</li>

In [60]:
mask = df['Mode of study'] == "All"

<ol start='3'>
<li>Use the variable named mask to update the dataframe named df so that the dataframe is filtered based on the boolean mask</li>
<li>Output the unique values for the column 'Mode of study' to ensure that 'Full-time' and 'Part-time' have been filtered out (don't exist in the dataframe)</li>

In [61]:
df = df[mask]


<ol start='5'>
<li>Output the dataframe</li>

In [62]:
df['Mode of study'].unique()
df

Academic Year Mode of study     Sex Country of HE provider  \
0          2013/14           All     All                    All   
1          2013/14           All     All                    All   
2          2013/14           All     All                    All   
3          2013/14           All     All                    All   
4          2013/14           All     All                    All   
5          2013/14           All     All                England   
6          2013/14           All     All                England   
7          2013/14           All     All                England   
8          2013/14           All     All                England   
9          2013/14           All     All                England   
10         2013/14           All     All       Northern Ireland   
11         2013/14           All     All       Northern Ireland   
12         2013/14           All     All       Northern Ireland   
13         2013/14           All     All       Northern Ireland   
14         2013/14           All     All       Northern Ireland   
15         2013/14           All     All               Scotland   
16         2013/14           All     All               Scotland   
17         2013/14           All     All               Scotland   
18         2013/14           All     All               Scotland   
19         2013/14           All     All               Scotland   
20         2013/14           All     All                  Wales   
21         2013/14           All     All                  Wales   
22         2013/14           All     All                  Wales   
23         2013/14           All     All                  Wales   
24         2013/14           All     All                  Wales   
25         2013/14           All  Female                    All   
26         2013/14           All  Female                    All   
27         2013/14           All  Female                    All   
28         2013/14           All  Female                    All   
29         2013/14           All  Female                    All   
...            ...           ...     ...                    ...   
1170       2017/18           All    Male               Scotland   
1171       2017/18           All    Male               Scotland   
1172       2017/18           All    Male               Scotland   
1173       2017/18           All    Male               Scotland   
1174       2017/18           All    Male               Scotland   
1175       2017/18           All    Male                  Wales   
1176       2017/18           All    Male                  Wales   
1177       2017/18           All    Male                  Wales   
1178       2017/18           All    Male                  Wales   
1179       2017/18           All    Male                  Wales   
1180       2017/18           All   Other                    All   
1181       2017/18           All   Other                    All   
1182       2017/18           All   Other                    All   
1183       2017/18           All   Other                    All   
1184       2017/18           All   Other                    All   
1185       2017/18           All   Other                England   
1186       2017/18           All   Other                England   
1187       2017/18           All   Other                England   
1188       2017/18           All   Other                England   
1189       2017/18           All   Other                England   
1190       2017/18           All   Other               Scotland   
1191       2017/18           All   Other               Scotland   
1192       2017/18           All   Other               Scotland   
1193       2017/18           All   Other               Scotland   
1194       2017/18           All   Other               Scotland   
1195       2017/18           All   Other                  Wales   
1196       2017/18           All   Other                  Wales   
1197       2017/18           All   Other            

<h2>Boolean Masks Continued</h2>

<p>Create a boolean mask and then filter the data based on the categorical data 'All' for 'Sex'</p>
<p>Then create a boolean mask and then filter the data based on the categorical data 'All' for 'Country of HE provider'</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 1380 rows to 25 rows.</p>

In [64]:
mask1 = df['Sex'] == "All"
df = df[mask1]

mask2 = df['Country of HE provider'] == "All"

df = df[mask2]
df

Academic Year Mode of study  Sex Country of HE provider  \
0          2013/14           All  All                    All   
1          2013/14           All  All                    All   
2          2013/14           All  All                    All   
3          2013/14           All  All                    All   
4          2013/14           All  All                    All   
275        2014/15           All  All                    All   
276        2014/15           All  All                    All   
277        2014/15           All  All                    All   
278        2014/15           All  All                    All   
279        2014/15           All  All                    All   
550        2015/16           All  All                    All   
551        2015/16           All  All                    All   
552        2015/16           All  All                    All   
553        2015/16           All  All                    All   
554        2015/16           All  All                    All   
825        2016/17           All  All                    All   
826        2016/17           All  All                    All   
827        2016/17           All  All                    All   
828        2016/17           All  All                    All   
829        2016/17           All  All                    All   
1105       2017/18           All  All                    All   
1106       2017/18           All  All                    All   
1107       2017/18           All  All                    All   
1108       2017/18           All  All                    All   
1109       2017/18           All  All                    All   

     Classification of First Degree  Number Percentage  
0                             First   79410        20%  
1                      Lower second   95965        24%  
2                        Third/Pass   21905         6%  
3                  Total classified  395605        NaN  
4                      Upper second  198320        50%  
275                           First   81640        22%  
276                    Lower second   85250        23%  
277                      Third/Pass   20340         5%  
278                Total classified  370905        NaN  
279                    Upper second  183680        50%  
550                           First   88890        24%  
551                    Lower second   81595        22%  
552                      Third/Pass   19280         5%  
553                Total classified  376330        NaN  
554                    Upper second  186565        50%  
825                           First  100945        26%  
826                    Lower second   79515        20%  
827                      Third/Pass   19025         5%  
828                Total classified  391875        NaN  
829                    Upper second  192395        49%  
1105                          First  110475        28%  
1106                   Lower second   76210        19%  
1107                     Third/Pass   17830         4%  
1108               Total classified  396940        NaN  
1109                   Upper second  192425        48%

<h2>More Boolean Masks</h2>

<p>We are not interested in rows that show the totals of classification for each year, filter these out.</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 25 rows to 20 rows.</p>

In [65]:
mask3 = df['Classification of First Degree'] != "Total classified"
df = df[mask3]
df

Academic Year Mode of study  Sex Country of HE provider  \
0          2013/14           All  All                    All   
1          2013/14           All  All                    All   
2          2013/14           All  All                    All   
4          2013/14           All  All                    All   
275        2014/15           All  All                    All   
276        2014/15           All  All                    All   
277        2014/15           All  All                    All   
279        2014/15           All  All                    All   
550        2015/16           All  All                    All   
551        2015/16           All  All                    All   
552        2015/16           All  All                    All   
554        2015/16           All  All                    All   
825        2016/17           All  All                    All   
826        2016/17           All  All                    All   
827        2016/17           All  All                    All   
829        2016/17           All  All                    All   
1105       2017/18           All  All                    All   
1106       2017/18           All  All                    All   
1107       2017/18           All  All                    All   
1109       2017/18           All  All                    All   

     Classification of First Degree  Number Percentage  
0                             First   79410        20%  
1                      Lower second   95965        24%  
2                        Third/Pass   21905         6%  
4                      Upper second  198320        50%  
275                           First   81640        22%  
276                    Lower second   85250        23%  
277                      Third/Pass   20340         5%  
279                    Upper second  183680        50%  
550                           First   88890        24%  
551                    Lower second   81595        22%  
552                      Third/Pass   19280         5%  
554                    Upper second  186565        50%  
825                           First  100945        26%  
826                    Lower second   79515        20%  
827                      Third/Pass   19025         5%  
829                    Upper second  192395        49%  
1105                          First  110475        28%  
1106                   Lower second   76210        19%  
1107                     Third/Pass   17830         4%  
1109                   Upper second  192425        48%

<h2>Dropping Columns</h2>
<p>Since columns 'Mode of study', 'Sex', and 'Country of HE provider' don\t provide us with any meaninful information (the all contain the data 'All'), drop these columns from the dataframe.</p>
<p>Output the dataframe to check the columns have been dropped successfully.</p>

In [68]:
df = df.drop(['Mode of study', 'Sex', 'Country of HE provider'], axis=1)
df

Academic Year Classification of First Degree  Number Percentage
0          2013/14                          First   79410        20%
1          2013/14                   Lower second   95965        24%
2          2013/14                     Third/Pass   21905         6%
4          2013/14                   Upper second  198320        50%
275        2014/15                          First   81640        22%
276        2014/15                   Lower second   85250        23%
277        2014/15                     Third/Pass   20340         5%
279        2014/15                   Upper second  183680        50%
550        2015/16                          First   88890        24%
551        2015/16                   Lower second   81595        22%
552        2015/16                     Third/Pass   19280         5%
554        2015/16                   Upper second  186565        50%
825        2016/17                          First  100945        26%
826        2016/17                   Lower second   79515        20%
827        2016/17                     Third/Pass   19025         5%
829        2016/17                   Upper second  192395        49%
1105       2017/18                          First  110475        28%
1106       2017/18                   Lower second   76210        19%
1107       2017/18                     Third/Pass   17830         4%
1109       2017/18                   Upper second  192425        48%

<h2>Checking the Data Type of a Column</h2>
<p>Check the type of data that is stored for one of the values in the percentage column.</p>

In [70]:
df.dtypes

Academic Year                     object
Classification of First Degree    object
Number                             int64
Percentage                        object
dtype: object

<h2>Cleaning and converting to an string data type to a number data type</h2>
<p>The percentage column is a string data type because of the % character.  Remove the % character and convert the values for this column to a numeric data type</p>

In [102]:
df
for i in range(0,len(df.index)):
    value = df.at[i, 'Percentage']
    print(value)
    if type(value) != int:
        newvalue = value.replace('%', '')

        newvalue = int(newvalue)
        print(newvalue)

        df = df.replace(to_replace = value, value = newvalue)
    


20


AttributeError: 'numpy.int64' object has no attribute 'replace'

In [104]:
df
df.replace({'Percentage': "26%"}, 26)

level_0  index Academic Year Classification of First Degree  Number  \
0         0      0       2013/14                          First   79410   
1         1      1       2013/14                   Lower second   95965   
2         2      2       2013/14                     Third/Pass   21905   
3         3      4       2013/14                   Upper second  198320   
4         4    275       2014/15                          First   81640   
5         5    276       2014/15                   Lower second   85250   
6         6    277       2014/15                     Third/Pass   20340   
7         7    279       2014/15                   Upper second  183680   
8         8    550       2015/16                          First   88890   
9         9    551       2015/16                   Lower second   81595   
10       10    552       2015/16                     Third/Pass   19280   
11       11    554       2015/16                   Upper second  186565   
12       12    825       2016/17                          First  100945   
13       13    826       2016/17                   Lower second   79515   
14       14    827       2016/17                     Third/Pass   19025   
15       15    829       2016/17                   Upper second  192395   
16       16   1105       2017/18                          First  110475   
17       17   1106       2017/18                   Lower second   76210   
18       18   1107       2017/18                     Third/Pass   17830   
19       19   1109       2017/18                   Upper second  192425   

    Percentage  
0           20  
1           24  
2            6  
3           50  
4           22  
5           23  
6            5  
7           50  
8           24  
9           22  
10           5  
11          50  
12          26  
13          20  
14           5  
15          49  
16          28  
17          19  
18           4  
19          48

In [105]:
df

level_0  index Academic Year Classification of First Degree  Number  \
0         0      0       2013/14                          First   79410   
1         1      1       2013/14                   Lower second   95965   
2         2      2       2013/14                     Third/Pass   21905   
3         3      4       2013/14                   Upper second  198320   
4         4    275       2014/15                          First   81640   
5         5    276       2014/15                   Lower second   85250   
6         6    277       2014/15                     Third/Pass   20340   
7         7    279       2014/15                   Upper second  183680   
8         8    550       2015/16                          First   88890   
9         9    551       2015/16                   Lower second   81595   
10       10    552       2015/16                     Third/Pass   19280   
11       11    554       2015/16                   Upper second  186565   
12       12    825       2016/17                          First  100945   
13       13    826       2016/17                   Lower second   79515   
14       14    827       2016/17                     Third/Pass   19025   
15       15    829       2016/17                   Upper second  192395   
16       16   1105       2017/18                          First  110475   
17       17   1106       2017/18                   Lower second   76210   
18       18   1107       2017/18                     Third/Pass   17830   
19       19   1109       2017/18                   Upper second  192425   

    Percentage  
0           20  
1           24  
2            6  
3           50  
4           22  
5           23  
6            5  
7           50  
8           24  
9           22  
10           5  
11          50  
12          26  
13          20  
14           5  
15          49  
16          28  
17          19  
18           4  
19          48

<h2>New Dataframes</h2>
<p>Often it is useful to create seperate smaller dataframes, create four new dataframes named df_first, df_upp_sec, df_low_sec and df_third, to seperate the different degree classifications.</p>

Example:
df_first would output:
<pre>
      Academic Year   Classification of First Degree	Number	 Percentage
0	    2013/14	       First	                    79410	  20%
275	  2014/15	       First	                    81640	  22%
550	  2015/16	       First	                    88890	  24%
825	  2016/17	       First	                    100945	 26%
1105	 2017/18	       First	                    110475	 28%
</pre>

In [106]:
mask = df['Classification of First Degree'] == "First"
df_first = df[mask]
df_first

mask = df['Classification of First Degree'] == "Upper second"
df_upp_sec = df[mask]
df_upp_sec

mask = df['Classification of First Degree'] == "Lower second"
df_low_sec = df[mask]
df_low_sec

mask = df['Classification of First Degree'] == "Third/Pass"
df_third = df[mask]
df_third

level_0  index Academic Year Classification of First Degree  Number  \
2         2      2       2013/14                     Third/Pass   21905   
6         6    277       2014/15                     Third/Pass   20340   
10       10    552       2015/16                     Third/Pass   19280   
14       14    827       2016/17                     Third/Pass   19025   
18       18   1107       2017/18                     Third/Pass   17830   

    Percentage  
2            6  
6            5  
10           5  
14           5  
18           4

<h2>Save the Dataframes to CSV files</h2>
<p>Save the dataframes df, df_first, df_upp_sec, df_low_sec and df_third to csv files with filesnames that match the dataframe name.

In [107]:
df_first.to_csv("df_first.csv", encoding='utf-8', index=False)

In [108]:
df_upp_sec.to_csv("df_upp_sec.csv", encoding='utf-8', index=False)
df_low_sec.to_csv("df_low_sec.csv", encoding='utf-8', index=False)
df_third.to_csv("df_third.csv", encoding='utf-8', index=False)

In [109]:
df.to_csv("df.csv", encoding='utf-8', index=False)